### PR TITLE
feat: supported type of vue@2.7+

### LINKED ISSUE #28591: Support Vue 2.7.16
### What would you like?

Hi!
I doing this https://github.com/vuejs/create-vue/issues/408

It uses `cypress`. 

So far it looks like it has some type of error

But, need wait https://github.com/vuejs/vue-test-utils/issues/2087 to be closed

### Why is this needed?

In `Vue 2` final days, we need to try to update his peripheral

### Other

_No response_

### PR DESCRIPTION
**Pull request by baboon-king**
- Closes #28591

### Additional details

Add support for vue@2.7+ types.

### Steps to test

none

### How has the user experience changed?

Better type support in Vue@2.7
### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
@baboon-king Looks like there's an error being thrown in our build job. 
Screenshot 2024-01-29 at 10 44 35 AM

**Comment by baboon-king** [comment]
> @baboon-king Looks like there's an error being thrown in our build job. Screenshot 2024-01-29 at 10 44 35 AM

All right. Let's me look look!

**Comment by baboon-king** [comment]
@jennifer-shehane 

After my examination.

TBH.
This is a problem that comes from upstream. In "umd.d.ts" he imported an unexported type.
import: https://github.com/vuejs/vue/blob/main/types/umd.d.ts#L3
unexported: https://github.com/vuejs/vue/blob/main/types/options.d.ts#L163
But https://github.com/vuejs/vue?tab=readme-ov-file#vue-2-has-reached-end-of-life lol.

So. I need more time to try and fix it.

**Comment by baboon-king** [comment]
more info
https://rollupjs.org/troubleshooting/#error-name-is-not-exported-by-module

**Comment by baboon-king** [comment]
@jennifer-shehane Hi！
Excuse me.

In a nutshell, this is because rollup-plugin-dts has [`respectExternal: true`](https://github.com/cypress-io/cypress/blob/develop/npm/mount-utils/create-rollup-entry.mjs#L75).
This option is false by default, do you know why you set respectExternal to true.
For npm/vue2 build tasks can I turn off this option, or is it possible?

> All external dependencies from node_modules are automatically excluded from bundling. This can be overridden using the respectExternal setting, but it is generally not recommended. While rollup of external @types generally works, it is not recommended.
https://github.com/Swatinem/rollup-plugin-dts?tab=readme-ov-file#what-to-expect

**Comment by jennifer-shehane** [comment]
@baboon-king I don't recall the context on that one.

**Comment by baboon-king** [comment]
> @baboon-king I don't recall the context on that one.

it's Ok!.
Can you review my last commit. it's fixed https://github.com/cypress-io/cypress/pull/28818#issuecomment-1914972360
image

**Comment by baboon-king** [comment]
Hi! @jennifer-shehane 

The branch has been rebase , it's clean.

File: npm/vue2/rollup.config.mjs [hunk 1]
```diff
@@ -3,6 +3,9 @@ import json from '@rollup/plugin-json'
 import replace from '@rollup/plugin-replace'
 
 const config = {
+  dtsOptions: {
+    respectExternal: false
```
Comment (AtofStryker) [inline npm/vue2/rollup.config.mjs]: was the issue that were were rolling up the vue `.d.ts` files from `2.6.12` and that was causing issues?

Comment (baboon-king) [reply npm/vue2/rollup.config.mjs]: No, the issue is not related to rolling up the Vue .d.ts files from 2.6.12. This problem occurs in versions[ "2.7+"](https://github.com/cypress-io/cypress/pull/28818/commits/89a83cbc8399627c894fc948e2eb7195e9f06108) instead.

Comment (AtofStryker) [reply npm/vue2/rollup.config.mjs]: it sounds like we were bundling the declaration types previously which sounded like it caused a collision. Would this change use the declaration files for whatever version of vue2 is installed?

Comment (baboon-king) [reply npm/vue2/rollup.config.mjs]: This won't affect Vue 2. In most cases, this option (respectExternal) doesn't need to be enabled, at least not for Vue as Vue itself already includes its own types. The documentation of rollup-plugin-dts also specifies that this option is disabled by default.

The main issue is that after upgrading to Vue 2.7+, the respectExternal option of rollup-plugin-dts was causing the npm/vue2 build task to fail. That's why I disabled this option for npm/vue2.

**Review by AtofStryker — COMMENTED** [review]
@baboon-king can you help me reproduce the type error you were seeing in vue `2.7.16`? I created a reproduction repository [here](https://github.com/AtofStryker/cypress-issue-28591) but I can't seem to produce a type error under test. Is there something I am missing or forgetting to do here?

**Comment by baboon-king** [comment]
> @baboon-king can you help me reproduce the type error you were seeing in vue `2.7.16`? I created a reproduction repository [here](https://github.com/AtofStryker/cypress-issue-28591) but I can't seem to produce a type error under test. Is there something I am missing or forgetting to do here?

@AtofStryker see [here](https://github.com/baboon-king/typescript-cypress-issue-28591).

I created a reproduction case here.
https://github.com/baboon-king/typescript-cypress-issue-28591
For Terminal
1. `pnpm install`
2. `pnpm run type-check`

For vscode
1. Enable the `Vue.vscode-typescript-vue-plugin` plugin.
2. Look at lines 5 and 9 of `HelloWorld.cy.ts`.

File: npm/mount-utils/create-rollup-entry.mjs [hunk 2]
```diff
@@ -15,6 +15,7 @@ export function createEntries (options) {
     input,
     config = {},
   } = options
+  const { dtsOptions = {} } = config
```
Comment (AtofStryker) [inline npm/mount-utils/create-rollup-entry.mjs]:
```suggestion
```

File: npm/mount-utils/create-rollup-entry.mjs [hunk 3]
```diff
@@ -15,6 +15,7 @@ export function createEntries (options) {
     input,
     config = {},
```
Comment (AtofStryker) [inline npm/mount-utils/create-rollup-entry.mjs:16]:
```suggestion
    config = {},
    dtsOptions = {},
```

File: npm/vue2/rollup.config.mjs [hunk 4]
```diff
@@ -3,6 +3,9 @@ import json from '@rollup/plugin-json'
...
+  },
```
Comment (AtofStryker) [inline npm/vue2/rollup.config.mjs]:
```suggestion
```

File: npm/vue2/rollup.config.mjs [hunk 5]
```diff
@@ -3,6 +3,9 @@ import json from '@rollup/plugin-json'
 import replace from '@rollup/plugin-replace'
 
 const config = {
```
Comment (AtofStryker) [inline npm/vue2/rollup.config.mjs:9]:
```suggestion
const dtsOptions = {
  respectExternal: false,
}

const config = {
```

File: npm/vue2/rollup.config.mjs [hunk 6]
```diff
@@ -3,6 +3,9 @@ import json from '@rollup/plugin-json'
...
+  dtsOptions: {
+    respectExternal: false
+  },
   external: [
     'vue',
   ],
```
Comment (AtofStryker) [inline npm/vue2/rollup.config.mjs]: also need to pass `dtsOptions` into `createEntries`

**Review by AtofStryker — CHANGES_REQUESTED** [review]
The current method of passing in `dtsOptions`  causes warnings in the build. I recommended a work around to at least avoid this

**Review by baboon-king — COMMENTED** [review]
right！MVP

@AtofStryker 
I think it's great, but do I need to review your changes?

**Comment by AtofStryker** [comment]
> right！MVP
> 
> @AtofStryker I think it's great, but do I need to review your changes?

@baboon-king currently these are just suggestions in the review, which you should be able to apply and test out!

**Comment by baboon-king** [comment]
> > right！MVP
> > @AtofStryker I think it's great, but do I need to review your changes?
> 
> @baboon-king currently these are just suggestions in the review, which you should be able to apply and test out!

Ok, I see. 
And
I've applied and tested it.
at latest commit

File: npm/vue2/rollup.config.mjs [hunk 7]
```diff
@@ -25,4 +29,4 @@ const config = {
   },
 }
 
-export default createEntries({ formats: ['es', 'cjs'], input: 'src/index.ts', config })
+export default createEntries({ formats: ['es', 'cjs'], input: 'src/index.ts', config, dtsOptions})
```
Comment (AtofStryker) [inline npm/vue2/rollup.config.mjs]:
```suggestion
export default createEntries({ formats: ['es', 'cjs'], input: 'src/index.ts', config, dtsOptions })
```

**Review by AtofStryker — APPROVED** [review]
This seems correct based on https://github.com/vuejs/vue/blob/main/CHANGELOG.md#typescript-changes. Though it does feel weird that `v3` types would live inside an installed version of `vue@2`. Tested and things seem appropriate on my end

**Review by mschile — APPROVED** [review]

**Pull request merged**